onnxruntime_training_go.go

```diff
@@ -0,0 +1,531 @@
+
+// Scalar is like a tensor but the underlying go slice is of length 1 and it has no dimension.
+// It can be used to store e.g. the loss from a training cycle.
+type Scalar[T TensorData] struct {
+	data     []T
+	dataSize uintptr
+	ortValue *C.OrtValue
+}
+
+func (s *Scalar[T]) GetShape() Shape {
+	return nil
+}
+
+func (s *Scalar[T]) ZeroContents() {
+	C.memset(unsafe.Pointer(&s.data[0]), 0, C.size_t(s.dataSize))
+}
+
+func (s *Scalar[T]) Destroy() error {
+	C.ReleaseOrtValue(s.ortValue)
+	s.ortValue = nil
+	s.data = nil
+	s.dataSize = 0
+	return nil
+}
+
+// GetData returns the undelying data for the scalar.
+// If you want to explicitly set the scalar's data, use Set.
+func (t *Scalar[T]) GetData() T {
+	return t.data[0]
+}
+
+// Set allows to explicitly set the underlying value for the scalar.
+func (t *Scalar[T]) Set(value T) {
+	t.data = []T{value}
+}
+
+func (t *Scalar[T]) DataType() C.ONNXTensorElementDataType {
+	return GetTensorElementDataType[T]()
+}
+
+func (t *Scalar[_]) GetInternals() *TensorInternalData {
+	return &TensorInternalData{
+		ortValue: t.ortValue,
+	}
+}
+
+// NewEmptyScalar creates a new scalar of type T.
+func NewEmptyScalar[T TensorData]() (*Scalar[T], error) {
+	var data T
+	return NewScalar(data)
+}
+
+// NewScalar creates a new scalar of type T backed by a value of type T.
+// Note that, differently from tensors, this is not a []T but just a value T.
+func NewScalar[T TensorData](data T) (*Scalar[T], error) {
+	if !IsInitialized() {
+		return nil, NotInitializedError
+	}
+
+	dataSlice := []T{data}
+	var ortValue *C.OrtValue
+	dataType := GetTensorElementDataType[T]()
+	dataSize := unsafe.Sizeof(dataSlice[0]) * uintptr(1)
+
+	status := C.CreateOrtTensorWithShape(unsafe.Pointer(&dataSlice[0]),
+		C.size_t(dataSize), nil, C.int64_t(0), ortMemoryInfo, dataType, &ortValue)
+	if status != nil {
+		return nil, statusToError(status)
+	}
+	toReturn := Scalar[T]{
+		data:     dataSlice,
+		dataSize: dataSize,
+		ortValue: ortValue,
+	}
+	return &toReturn, nil
+}
+
+// TraininSession is the type that wraps the C training session object.
+type TrainingSession struct {
+	ortTrainingSession *C.OrtTrainingSession
+	ortCheckpointState *C.OrtCheckpointState
+	inputs             []*C.OrtValue
+	outputs            []*C.OrtValue
+	trainingModelPath  *C.char
+	optimizerModelPath *C.char
+	evalModelPath      *C.char
+}
+
+// ExportModel is used to export the final trained model to disk. It requires the path for
+// the exported model as well as the names of the graph nodes to export.
+// Note that currently the final model can only be exported if the session has been
+// initialized with NewTrainingSession and the path to the eval model has been provided.
+func (s *TrainingSession) ExportModel(path string, outputNames []string) error {
+	if s.evalModelPath == nil {
+		return fmt.Errorf(`final model can only be exported if the eval model path is 
```

Comment:
Preferably splice together this long string using `+`, i.e.,
```go
fmt.Errorf("first part " +
    "second part")
```
since the backtick will include the newline in the error message.